pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -159,13 +170,67 @@ func (s *Scheduler) add(pod *v1.Pod) error {
 	// Create new node
 	var errs error
 	for _, nodeTemplate := range s.nodeTemplates {
-		node := NewNode(nodeTemplate, s.topology, s.daemonOverhead[nodeTemplate], s.instanceTypes)
+		instanceTypes := s.instanceTypes
+		// if limits have been applied to the provisioner, ensure we filter instance types to avoid violating those limits
+		if remaining, ok := s.remainingResources[nodeTemplate.ProvisionerName]; ok {
+			instanceTypes = filterByRemainingResources(s.instanceTypes, remaining)
+			if len(instanceTypes) == 0 {
+				errs = multierr.Append(errs, fmt.Errorf("all available instance types exceed provisioner limits"))
+				continue
+			}
+		}
+
+		node := NewNode(nodeTemplate, s.topology, s.daemonOverhead[nodeTemplate], instanceTypes)
 		err := node.Add(pod)
 		if err == nil {
 			s.nodes = append(s.nodes, node)
+			// we will launch this node and need to track its maximum possible resource usage against our remaining resources
+			s.remainingResources[nodeTemplate.ProvisionerName] = subtractMax(s.remainingResources[nodeTemplate.ProvisionerName], node.InstanceTypeOptions)
 			return nil
 		}
 		errs = multierr.Append(errs, err)
 	}
 	return errs
 }
+
+// subtractMax returns the remaining resources after subtracting the max resource quantity per instance type. To avoid
+// overshooting out, we need to pessimistically assume that if e.g. we request a 2, 4 or 8 CPU instance type
+// that the 8 CPU instance type is all that will be available.  This could cause a batch of pods to take multiple rounds
+// to schedule.
+func subtractMax(remaining v1.ResourceList, instanceTypes []cloudprovider.InstanceType) v1.ResourceList {
```

Comment:
There are a couple of similar helpers in cloudprovider/requirements.go that filter instance types based off of requirements. Thoughts on them living together?